### ISSUE TITLE
Enable analysis concurrent execution

### ISSUE THREAD

**Issue by costin-zaharia-sonarsource** [SonarSource/sonar-dotnet#3164]
Currently the analysis is done sequentially but the build time can be significantly reduced by enabling concurrent execution.

On a proof a concept, analyzing nodatime, we reduced the build time from  **~0:01:07** to **~0:00:36**.

**Comment by Evangelink** [comment]
@costin-zaharia-sonarsource Be careful when activating this feature as some of the rules where not written with the multi-threaded environment in mind.

**Comment by costin-zaharia-sonarsource** [comment]
> @costin-zaharia-sonarsource Be careful when activating this feature as some of the rules where not written with the multi-threaded environment in mind.

Hi @Evangelink, 

indeed in the tests I've done I noticed errors and unexpected results, a lot of testing will be needed if we decide to go for it. Shared static state (https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/Common/SymbolicValues/SymbolicValue.cs#L92) is the first thing which comes into my mind regarding this but there might be other problems too. 

Regarding the performance improvements, they were not consistent and depended a lot on the project being analyzed so we have to see if it's worth the change considering the effort.

**Comment by Evangelink** [comment]
@costin-zaharia-sonarsource When we looked at this, the best solution that came out was to change a bit the infrastructure to allow each rule to specify whether it can be parallelized or not. This way you can progressively enable rules when you have done enough tests to ensure the results. BTW if you still have the dogfood infrastructure it might be a good way to see if you have some diffs for performance and also reported issues.

Clearly all dataflow rules shall remain sequential. Note that for future, you might want to revisit the dataflow by relying on the one provided in roslyn.

**Comment by costin-zaharia-sonarsource** [comment]
> the best solution that came out was to change a bit the infrastructure to allow each rule to specify whether it can be parallelized or not

This approach sounds very good indeed.

> Note that for future, you might want to revisit the dataflow by relying on the one provided in roslyn.

Yes, we have this in mind already. Our plan is to implement an existing rule with Roslyn CFG and dataflow engine and compare the results. Once this is done we could take a better decision.

**Comment by costin-zaharia-sonarsource** [comment]
Copying here some notes I've made in February when I've did a POC: 
- Build time was reduced for nodatime from ~0:01:07 to ~0:00:36 but on openiddict-core I didn’t see the same improvement. I think we need to investigate this further since for some projects we could get a significant improvement but this doesn’t seem consistent
- I’ve tested a build with concurrent execution enabled and for nodatime, n2cms and openiddict-core the analysis works fine (outputs the same warnings) but on efcore the symbolic execution rules randomly fail at SonarAnalyzer.SymbolicExecution.CSharpExplodedGraph.EnsureStackState(ExpressionSyntax parenthesizedExpression, ProgramState programState). As far as I can tell `MemberShouldNotHaveConflictingTransparencyAttributes` and all the symbolic execution rules need to be modified to be able to run concurrently. https://github.com/SonarSource/sonar-dotnet/issues/3164

Also, we need to take a bit of time and run the performance analysis again. I expect that numbers will be different now that all symbolic execution rules are executed in a single pass. An option would be to do this test while enabling concurrent execution for all rules except symbolic execution.
